### ISSUE TITLE
binary_cross_entropy (loss) seems to be giving incorrect values for very negative logits

### ISSUE THREAD

**Issue by dhruvbird** [pytorch/pytorch#103316]
### 🐛 Describe the bug

Please see the following code which computes the loss manually (using PyTorch ops) and using the `F.binary_cross_entropy` API.

```
import torch
from torch import nn
from torch.nn import functional as F

x = nn.Parameter(torch.tensor([-50., -10., -5., -2., 0., 2., 5., 10.]), requires_grad=True)

optimizer = torch.optim.Adam([x])
optimizer.zero_grad()

xs = x.sigmoid()
loss = -xs.log().mean()

loss.backward()
print(x.grad)

optimizer.zero_grad()
loss = F.binary_cross_entropy(x.sigmoid(), torch.ones_like(x))
loss.backward()
print(x.grad)
```

This code prints.

```
tensor([-1.2500e-01, -1.2499e-01, -1.2416e-01, -1.1010e-01, -6.2500e-02,
        -1.4900e-02, -8.3660e-04, -5.6773e-06])
tensor([-2.4109e-11, -1.2499e-01, -1.2416e-01, -1.1010e-01, -6.2500e-02,
        -1.4900e-02, -8.3660e-04, -5.6773e-06])
```

Note that for the input `-50`, the `binary_cross_entropy` API returns a very small negative value `-2.4109e-11`, instead of a larger negative value, namely `-1.2500e-01`.

### Versions

1.12.1

**Comment by Potato-Cracker** [comment]
Hi, good catch. However, this behavior is expected due to numerical stability reasons. You can about this in the description of BCELoss [here](https://github.com/pytorch/pytorch/blob/1c3a7d9a7ef1d330d4d32ed2dcd8aab92af23abc/torch/nn/modules/loss.py#L538).

**Comment by dhruvbird** [comment]
> Hi, good catch. However, this behavior is expected due to numerical stability reasons. You can about this in the description of BCELoss [here](https://github.com/pytorch/pytorch/blob/1c3a7d9a7ef1d330d4d32ed2dcd8aab92af23abc/torch/nn/modules/loss.py#L538).

I'm not sure if it's related to the numerical instability issues arising from that, since I have manually implemented the BCE loss above and am feeding it the result of `sigmoid`, and it seems to be giving a better result.

I also verified that the same issue occurs on 2.x

**Comment by Potato-Cracker** [comment]
Your manual loss calculation seems correct. However, from what I've read and worked on, pytorch prevents loss from becoming too large to avoid the exploding gradient problem.
This was what was given in the docs-

>     Notice that if :math:`x_n` is either 0 or 1, one of the log terms would be
    mathematically undefined in the above loss equation. PyTorch chooses to set
    :math:`\log (0) = -\infty`, since :math:`\lim_{x\to 0} \log (x) = -\infty`.
    However, an infinite term in the loss equation is not desirable for several reasons.

    For one, if either :math:`y_n = 0` or :math:`(1 - y_n) = 0`, then we would be
    multiplying 0 with infinity. Secondly, if we have an infinite loss value, then
    we would also have an infinite term in our gradient, since
    :math:`\lim_{x\to 0} \frac{d}{dx} \log (x) = \infty`.
    This would make BCELoss's backward method nonlinear with respect to :math:`x_n`,
    and using it for things like linear regression would not be straight-forward.

    Our solution is that BCELoss clamps its log function outputs to be greater than
    or equal to -100. This way, we can always have a finite loss value and a linear
    backward method.

You can see the clamping of log function being done here- https://github.com/pytorch/pytorch/blob/1c3a7d9a7ef1d330d4d32ed2dcd8aab92af23abc/torch/_decomp/decompositions.py#L570

Further,
`BCELoss = - y * log(x) - (1 - y) * log(1 - x)
`

The first term in the loss, -y * log(x), is equivalent to -log(x) because y is 1. Second term becomes 0 cause 1-y = 0

When you manually compute the sigmoid of -50 (which is effectively zero due to the nature of the sigmoid function), and then take the logarithm, you get -inf due to the nature of the log function. However, PyTorch's backward() function can't handle -inf, so it replaces it with a very large negative number. This is likely what's causing the -1.2500e-01 gradient. I am not well versed with how backward() works in pytorch.

On the other hand, when you use binary_cross_entropy, PyTorch applies some numerical stability tricks to prevent the -inf. This is why you're seeing a gradient close to 0 (-2.4109e-11) for -50 when using binary_cross_entropy.
Correct me if I am wrong :)

**Comment by dhruvbird** [comment]
> When you manually compute the sigmoid of -50 (which is effectively zero due to the nature of the sigmoid function), and then take the logarithm, you get -inf due to the nature of the log function. However, PyTorch's backward() function can't handle -inf, so it replaces it with a very large negative number. This is likely what's causing the -1.2500e-01 gradient. I am not well versed with how backward() works in pytorch.

Sigmoid(-50) is correctly computed by PyTorch as `1.9287e-22` (and not -inf), and then the log of that is correctly computed as -50.

```
>>> import torch
>>> x = torch.tensor([-50])
>>> print(x.sigmoid(), x.sigmoid().log())

tensor([1.9287e-22]) tensor([-50.])
```

**Comment by soulitzer** [comment]
cc @janeyx99
